### DISCUSSION TITLE
Symbol-Visibillity

### DISCUSSION THREAD

**Discussion by Vollstrecker** [pupnp/pupnp#372] [Ideas]
As mentioned in #365 atm. all symbols are exported, so also the internal stuff not meant for public use is accessible for others.
This is done by gcc by default, on windows it's manually turned on as otherwise nothing would be exported and the lib would be useless. This is worked around bei libtool by examining the compiled objects, building a version-script with declarations what should be exported which is then fed to the linker.

As cmake isn't able to insert steps between compiling and linking (in fact it does by providing a custom linker-command but this will be really ugly and maybe has side-effects when chainbuilding) we only can replicate this by using the wellknown __decspec(dllexport) on windows and __attribute__(visibillity.. on the others (didn't check how it works on cygwin). As I don't have the time atm. I just want to gather opinions in which way we want to get this done.

1. Add a definition UPNP_EXPORT with _declspec or __attribute depending on the system to a not installed config-header and add the -fvisibillity=hidden flags to the linker-command.
2. In addition add UPNP_INTERNAL which specifies __attribute to hidden (and nothing on win) and not needing another linker-flag.
3. Manually maintain a list of api-functions to export and tell each compiler what it should export and what not.
4. Leave everything as is.

I guess 4. isn't really an option, as this influences both size and loading-times of the lib. Not to mention that it enables other projects to use internal functions that are not meant for them, so it would break their build if we see a function is not needed anymore and remove it but they rely on it.

Option 3 will be a hell to maintain plus mostly unreadable scripts to generate the right input for the compilers. Maybe we also hit a commandline lenght limit as msvc needs that as params.

From the remaining 2 options I would prefer the second one. A little bit more work but it's clear what was intended and that the func was not just forgotten to check.

After that we could check if libtool is still required, or if can be dropped so build-deps are lowered to just plain autotools.

**Comment by whyman** [comment]
There is `EXPORT_SPEC` already, so I guess that just needs wiring up?

I think option 1 is the simplest.

**Reply by Vollstrecker** [reply]
Saw that, but this name is to generic and could bleed out to other projects. I'm thinking more about something like moving that into autoconfig.h (iirc this is the internal config) and then using the renamed spec with #ifndef if the dllimport statement is really needed (not sure if it needs to be mentioned explecit).

And surely option 1 is the simplest, that's why I mentioned option 2 is more work. Otoh. it could make clear that this is not for public use and maybe while doing this 1 or 2 old functions are found that just aren't used anywhere anymore.

**Reply by Vollstrecker** [reply]
Addendum: Just read about the dllimport, it is optional an can be omitted for functions. The same effect can be achieved when comiling the importing program with /GL option which we could enforce via interface. This doesn't work when using different Visual Studio versions for the lib and the app but as we don't ship binaries I guess it's save to assume that everything get's built on the same machine.

**Reply by mrjimenez** [reply]
Well, like @whyman  said, `EXPORT_SPEC` was supposed to do so. The name can be changed it that is de case.

In principle, the public interface is what is exposed by upnp.h. All the rest is internal, or at lease, should be. I like option 1 too.

**Reply by Vollstrecker** [reply]
k, atm I'm playing with doxygen, after that I'll look at this.
